core/src/main/java/org/apache/iceberg/ManifestEntry.java

```diff
@@ -118,6 +124,27 @@ default boolean isLive() {
   @Deprecated
   void setSequenceNumber(long sequenceNumber);
 
+  /**
+   * Returns the file sequence number.
+   *
+   * <p>The file sequence number represents the sequence number of the snapshot in which the
+   * underlying file was added. The file sequence number is always assigned at commit and cannot be
+   * provided explicitly, unlike the data sequence number. The file sequence number does not change
+   * upon assigning and must be preserved in existing and deleted entries.
+   *
+   * <p>This method can return null if the file sequence number is unknown. This may happen while
+   * reading a v2 manifest that did not persist the file sequence number for manifest entries with
+   * status EXISTING or DELETED (older Iceberg versions).
+   */
+  Long fileSequenceNumber();
```

Comment:
Do we have cases that only `fileSequenceNumber` cannot handle? I 'm wondering Can we replace `dataSequenceNumber` with `fileSequenceNumber` eventually.

Reply:
I think we need both as they serve different purposes. The data sequence number will still be used to decide whether a particular delete should apply. The file sequence number, on the other hand, will always tell us the sequence number of the snapshot that added this file. These values can be different during compaction as we may add a file with an old data sequence number.

Reply:
Oh, I see. Deletes should not apply to files that have larger `dataSequenceNumber` but smaller `fileSequenceNumber`.  I guess when compacting data files with deletes may lead to adding existing files with old `fileSequenceNumber`, compacting v1 table should not have that case, right?

Reply:
We already use `dataSequenceNumber` / `sequenceNumber` in scan planning. We just renamed how it was referenced in core recently. If we choose to expose these on `ContentFile` that's fine, but something we should do in a separate PR.

Reply:
Correct, these sequence numbers would serve different purposes and may be added to `ContentFile` later.

Reply:
@flyrain, this is something controlled by the following option in the compaction action.

```
  /**
   * If the compaction should use the sequence number of the snapshot at compaction start time for
   * new data files, instead of using the sequence number of the newly produced snapshot.
   *
   * <p>This avoids commit conflicts with updates that add newer equality deletes at a higher
   * sequence number.
   *
   * <p>Defaults to true.
   */
  String USE_STARTING_SEQUENCE_NUMBER = "use-starting-sequence-number";
```